runtime/memprof.c

```diff
@@ -248,100 +268,111 @@ static value capture_callstack_major()
   return res;
 }
 
-static value capture_callstack_minor()
+static value capture_callstack_minor(void)
 {
   value res;
   intnat wosize = caml_current_callstack_size(callstack_size);
-  CAMLassert(caml_memprof_suspended); /* => no samples in the call stack. */
+  CAMLassert(!caml_memprof_suspended);
+  caml_memprof_suspended = 1; /* => no samples in the call stack. */
   res = caml_alloc(wosize, 0);
+  caml_memprof_suspended = 0;
   caml_current_callstack_write(res);
   return res;
 }
 
+/**** Handling postponed sampled blocks. ****/
+/* When allocating in from C code, we cannot call the callback,
+   because the [caml_alloc_***] are guaranteed not to do so. These
+   functions make it possible to register a sampled block in a
+   todo-list so that the callback call is performed when possible. */
+/* Note: the shorter the delay is, the better, because the block is
+   linked to a root during the delay, so that the reachability
+   properties of the sampled block are artificially modified. */
+
 struct caml_memprof_postponed_block {
   value block;
   value callstack;
   int32_t occurrences;
+  enum ml_alloc_kind kind;
   struct caml_memprof_postponed_block* next;
-} *caml_memprof_postponed_head = NULL;
+} *caml_memprof_postponed_hd = NULL, *caml_memprof_postponed_tl = NULL;
 
-/* When allocating in the major heap, we cannot call the callback,
-   because [caml_alloc_shr] is guaranteed not to call the GC. Hence,
-   this function determines if the block needs to be sampled, and if
-   so, it registers the block in the todo-list so that the callback
-   call is performed when possible. */
-void caml_memprof_track_alloc_shr(value block)
-{
-  int32_t occurrences;
-  CAMLassert(Is_in_heap(block));
-  /* This test also makes sure memprof is initialized. */
-  if(lambda == 0 || caml_memprof_suspended)
-    return;
-  occurrences = mt_generate_poisson(Whsize_val(block));
-  if(occurrences > 0) {
-    struct caml_memprof_postponed_block* pb =
+static void free_postponed_block(struct caml_memprof_postponed_block *p) {
+  caml_remove_generational_global_root(&p->callstack);
+  caml_remove_generational_global_root(&p->block);
+  caml_stat_free(p);
+}
```

Comment:
(My comment below is not specifically for the present PR, it's fine to do nothing about it, but I thought it could be discussed.)

Upon seeing all this linked-list code again, I wondered if it would be possible to reuse a more generic mechanism. There are no generic linked lists that I can find in the runtime codebase, but there is the `generic_table` mechanism for dynamic arrays in `minor_gc.c`. Wouldn't it be nice to reuse them here? (One thing that is a nice with generic arrays is that they tend to not make many small `malloc` calls but a few larger calls).

If we wanted to use generic tables, we would have to be a bit careful in the `handle_postponed_blocks` function in case one callback fails with an exception. The generic-table operation that I see in `minor_gc.c` always handle the whole table at once, and then `clear` it wholesale. Here you would need to handle blocks starting from the end, releasing them one-by-one (and decreasing the current-element pointer), and stop in the exception case without clearing the un-handled entries. (Thinking about it, this implements LIFO behavior, which is not as nice as FIFO behavior for delayed callbacks...)

Reply:
Not sure this answers your concern, but I finally changed the linked list FIFO with a circular buffer. Now that I think of reusing the postponing system for sampling while unmarshalling (PR to come...), this system is going to be used extensively. I am not so sure that performances are so important here, but clearly a circular buffer is going to be more efficient than a linked list implemented with `malloc` and `free`.